crates/hotshot/task-impls/src/stats.rs

```diff
@@ -0,0 +1,346 @@
+                self.replica_entry(proposal.data.view_number())
+                    .proposal_validated = Some(now);
+            },
+            HotShotEvent::DaProposalSend(proposal, _) => {
+                self.leader_entry(proposal.data.view_number())
+                    .da_proposal_send = Some(now);
+            },
+            HotShotEvent::DaVoteSend(simple_vote) => {
+                self.replica_entry(simple_vote.view_number()).vote_send = Some(now);
+            },
+            HotShotEvent::QcFormed(either) => {
+                match either {
+                    Either::Left(qc) => {
+                        self.leader_entry(qc.view_number() + 1).qc_formed = Some(now)
+                    },
+                    Either::Right(tc) => {
+                        self.leader_entry(tc.view_number())
+                            .timeout_certificate_formed = Some(now)
+                    },
+                };
+            },
+            HotShotEvent::Qc2Formed(either) => {
+                match either {
+                    Either::Left(qc) => {
+                        self.leader_entry(qc.view_number() + 1).qc_formed = Some(now)
+                    },
+                    Either::Right(tc) => {
+                        self.leader_entry(tc.view_number())
+                            .timeout_certificate_formed = Some(now)
+                    },
+                };
+            },
+            HotShotEvent::DacSend(simple_certificate, _) => {
+                self.leader_entry(simple_certificate.view_number())
+                    .da_cert_send = Some(now);
+            },
+            HotShotEvent::ViewChange(view, epoch) => {
+                // Record the timestamp of the first observed view change
+                // This can happen when transitioning to the next view, either due to voting
+                // or receiving a proposal, but we only store the first one
+                if self.replica_entry(*view + 1).view_change.is_none() {
+                    self.replica_entry(*view + 1).view_change = Some(now);
+                }
+
+                if *epoch <= self.epoch && *view <= self.view {
+                    return Ok(());
+                }
+                if self.view < *view {
+                    self.view = *view;
+                }
+                let mut new_epoch = false;
+                if self.epoch < *epoch {
+                    self.epoch = *epoch;
+                    new_epoch = true;
+                }
+                if *view == TYPES::View::new(0) {
+                    return Ok(());
+                }
+
+                if new_epoch {
+                    let _ = self.dump_stats();
+                    self.garbage_collect(*view - 1);
+                }
+
+                let leader = self
+                    .membership_coordinator
+                    .membership_for_epoch(*epoch)
+                    .await?
+                    .leader(*view)
+                    .await?;
```

Comment:
Not sure if that matters but we have 2 async calls here and we take the timestamp at the start of the function, so the timestamp set as builder_start might be a bit stale.

Reply:
I think we want to take the stat time before we do the async calls on purpose.  The stats task isn't doing the actual processing so if it takes a long time we don't want to make it seem like the Hotshot did something faster.